Add `strict_override_decorator` option (PEP 698)

cdce8p opened the pull request:
Add the strict mode for [PEP 698](https://peps.python.org/pep-0698/#strict-enforcement-per-project).

Closes: #14072

ethanhs commented on mypy/messages.py:
```diff
@@ -1521,6 +1521,13 @@ def no_overridable_method(self, name: str, context: Context) -> None:
 
+    def override_decorator_missing(self, name: str, context: Context) -> None:
+        self.fail(
+            f'Method "{name}" is not marked as override '
+            "but is overriding a method in a base class",
```
I think saying it is "not marked as override" might be a bit confusing, perhaps this could be rephrased slightly like so?
```suggestion
            f'Method "{name}" is not marked by an override decorator'
            "but is overriding a method in a base class",
```

cdce8p replied:
I actually tried to copy the error message `pyright` uses. It's something like
```
Method "method1" is not marked as override but is overriding a method in class "Base"
```

_Note that pyright explicity mentions the base class here, whereas mypy does not. If that is useful, it could always be added later._

ethanhs reviewed (approved):
One minor wording suggestion but I think this looks good otherwise!

tmke8 commented on test-data/unit/check-functions.test:
```diff
@@ -3007,3 +3007,75 @@ class C(A):
+
+class C(A):
+    def f(self, y: int) -> str: pass  # E: Method "f" is not marked as override but is overriding a method in a base class
+[typing fixtures/typing-override.pyi]
```
Should this new fixture also be used for the existing override tests above? (Which I wrote, and where I just used typing-full...)

cdce8p commented:
Pushed a new update. The error message now includes the `base` class name. Similar to pyright.
The PR would be ready now.

tmke8 commented on mypy/checker.py:
```diff
@@ -643,9 +643,17 @@ def _visit_overloaded_func_def(self, defn: OverloadedFuncDef) -> None:
-            found_base_method = self.check_method_override(defn)
-            if defn.is_explicit_override and found_base_method is False:
+            found_method_base_classes = self.check_method_override(defn)
+            if defn.is_explicit_override and not found_method_base_classes:
```
`not found_method_base_class` also is `True` if `found_method_base_classes` is `None`, which is not what we want, I think. Because `None` indicates that another pass has to be made.

```suggestion
            if defn.is_explicit_override and found_method_base_classes is not None and not found_method_base_classes:
```

tmke8 replied:
I think you can do this more elegantly by changing the order of the if-branches.

JukkaL commented on docs/source/class_basics.rst:
```diff
@@ -233,6 +233,12 @@ show an error:
+
+   Use ``--strict-override-decorator`` or
+   :confval:`strict_override_decorator = True <strict_override_decorator>` to require
+   methods overrides use the ``@override`` decorator. Emit an error if it is missing.
```
"methods overrides use" -> "that method overrides use"?

Mention above that the decorator is included in 3.12 and must be imported from `typing_extensions` if using an earlier Python version, in particular since 3.12 is not out yet.

JukkaL commented on mypy/checker.py:
```diff
@@ -4739,9 +4762,21 @@ def visit_decorator(self, e: Decorator) -> None:
+            elif (
+                found_method_base_classes
+                and self.options.strict_override_decorator
+                and not e.func.is_explicit_override
```
Similar to above -- what about `__init__` and `__new__`? I wonder if the logic could be shared in a helper, at least partially, since we have similar code in three places.

JukkaL commented on mypy/main.py:
```diff
@@ -824,6 +824,14 @@ def add_invertible_flag(
+        "--strict-override-decorator",
+        default=False,
+        strict_flag=False,
+        help="Require override decorator if method is overriding a base class method.",
```
Nit: Please drop the period from the end of the help message, for consistency.

JukkaL commented on test-data/unit/check-functions.test:
```diff
@@ -3012,5 +3006,122 @@ class C(A):
+    def f(self, y: int) -> str: ...
+    @overload
+    def f(self, y: str) -> str: ...
+    @override
```
Test what happens if `@override` is used in one of the overload signatures before the implementation.

JukkaL commented on test-data/unit/check-functions.test:
```diff
@@ -3012,5 +3006,122 @@ class C(A):
+
+[case requireExplicitOverrideMultipleInheritance]
+# flags: --strict-override-decorator --python-version 3.12
+from typing import override
```
Test using `override` imported from `typing_extensions` on an older Python version?

cdce8p replied:
I rather not do that. Since none of our test fixtures use `@override`, we inadvertently get errors for those as well. That's also why I added a separate fixture for these test cases with just enough definitions and no accidental overrides. I would have to do the same just for the `typing_extensions` fixture.

There is also a test case for an unused `@override` already. As this PR doesn't change the decorator parsing, it probably fine to skip it.
https://github.com/python/mypy/blob/67cc05926b037243b71e857a394318c3a09d1daf/test-data/unit/check-functions.test#L2937-L2947

JukkaL commented on docs/source/config_file.rst:
```diff
@@ -714,6 +714,13 @@ section of the command line docs.
+    :type: boolean
+    :default: False
+
+    Require ``override`` decorator if method is overriding a base class method.
```
Maybe write "Require `@override` decorator ..." to make things a bit more explicit, since override is a common word.

JukkaL reviewed (commented):
Thanks for the PR! Looks good overall, some suggestions below. This is an important feature to have in mypy.

cdce8p commented:
Thanks for the review @JukkaL! Using a new error code instead of a cmd line option does make sense. The change was fairly strait forward. With that I also updated the documentation and added a whole new section in `error_code_list2.rst`.

For now I used `explicit-override` as error code. Happy to change that if we can come up with something better though.

JukkaL reviewed (approved):
Thanks for the updates, looks good!
